src/block.js

```diff
@@ -0,0 +1,201 @@
+const links = function * (decoded, path = []) {
+  if (typeof decoded !== 'object' || !decoded) return
+  for (const key of Object.keys(decoded)) {
+    const _path = path.slice()
+    _path.push(key)
+    const val = decoded[key]
+    if (val && typeof val === 'object') {
+      if (Array.isArray(val)) {
+        for (let i = 0; i < val.length; i++) {
+          const __path = _path.slice()
+          __path.push(i)
+          const o = val[i]
+          const cid = CID.asCID(o)
+          if (cid) {
+            yield [__path.join('/'), cid]
+          } else if (typeof o === 'object') {
+            yield * links(o, __path)
+          }
+        }
+      } else {
+        const cid = CID.asCID(val)
+        if (cid) {
+          yield [_path.join('/'), cid]
+        } else {
+          yield * links(val, _path)
+        }
+      }
+    }
+  }
+}
+
+const tree = function * (decoded, path = []) {
+  if (typeof decoded !== 'object' || !decoded) return
+  for (const key of Object.keys(decoded)) {
+    const _path = path.slice()
+    _path.push(key)
+    yield _path.join('/')
+    const val = decoded[key]
+    if (val && typeof val === 'object' && !CID.asCID(val)) {
+      if (Array.isArray(val)) {
+        for (let i = 0; i < val.length; i++) {
+          const __path = _path.slice()
+          __path.push(i)
+          const o = val[i]
+          yield __path.join('/')
+          if (typeof o === 'object' && !CID.asCID(o)) {
+            yield * tree(o, __path)
+          }
+        }
+      } else {
+        yield * tree(val, _path)
+      }
+    }
+  }
+}
+/* eslint-enable max-depth */
+
+class Block {
+  constructor ({ codec, hasher, source, cid, data }) {
+    setImmutable(this, '_codec', codec)
+    setImmutable(this, '_hasher', hasher)
+    if (source) setImmutable(this, '_source', source)
+    if (cid) setImmutable(this, '_cid', cid)
+    if (data) setImmutable(this, '_data', data)
+    setImmutable(this, 'asBlock', this)
+  }
+
+  decodeUnsafe () {
+    if (typeof this._source !== 'undefined') return this._source
+    throw new Error('Block created without a decoded state')
+  }
+
+  decode () {
+    const decoded = this.decodeUnsafe()
+    if (decoded === null) return null
+    if (isBinary(decoded)) return copyBinary(decoded)
+    if (immutableTypes.has(typeof decoded) || decoded === null) {
+      return decoded
+    }
+    return clone(decoded)
+  }
+
+  encodeUnsafe () {
+    if (this._data) return this._data
+    throw new Error('Block created without an encoded state')
+  }
+
+  encode () {
+    const data = this.encodeUnsafe()
+    return copyBinary(data)
+  }
+
+  async cid () {
+    if (this._cid) return this._cid
+    const hash = await this._hasher.digest(this.encodeUnsafe())
+    const cid = CID.create(1, this._codec.code, hash)
+    setImmutable(this, '_cid', cid)
+    return cid
+  }
+
+  get code () {
+    if (this._cid) return this._cid.code
+    return this._codec.code
+  }
+
+  async equals (block) {
```

Comment:
Nit: Might be worth renaming this. When I see `.equals` I assume a sync function.